### PR TITLE
[#826] switch to source layout

### PR DESCRIPTION
**Pull request by gaborbernat**
a POC of it at least

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -115,4 +115,4 @@ include_trailing_comma = True
 force_grid_wrap = 0
 line_length = 99
 known_first_party = tox
-known_third_party = pkg_resources,pluggy,py,pytest,setuptools,six
```
Comment (gaborbernat) [inline tox.ini]: @asottile I think this is a false positive of the isort plugin

Comment (asottile) [reply tox.ini]: yeah, it assumes a specific layout.  PRs welcome, the missing bits are to set the application directories:

For example, here's how `reorder_python_imports` handles this:
https://github.com/asottile/reorder_python_imports/blob/794f0498340ab3adbef443431993f8a8b2046da7/reorder_python_imports.py#L405-L411

You'll want:

```yaml
    args: [--application-directories, src]
```

once `seed-isort-config` supports that

Comment (gaborbernat) [reply tox.ini]: I'm caught up now with other stuff, so if you can fix it that's cool :+1:

**Comment by asottile** [comment]
Here ya go:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 8a1d348..3469cf6 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -6,9 +6,10 @@ repos:
       args: [--line-length=99, --safe]
       python_version: python3.6
 -   repo: https://github.com/asottile/seed-isort-config
-    rev: v0.1.0
+    rev: v1.0.0
     hooks:
     -   id: seed-isort-config
+        args: [--application-directories, src]
 -   repo: https://github.com/pre-commit/mirrors-isort
     rev: v4.3.4
     hooks:
diff --git a/tox.ini b/tox.ini
index 548f602..9acd122 100644
--- a/tox.ini
+++ b/tox.ini
@@ -115,4 +115,4 @@ include_trailing_comma = True
 force_grid_wrap = 0
 line_length = 99
 known_first_party = tox
-known_third_party = pkg_resources,pluggy,py,pytest,setuptools,six,tox
+known_third_party = pkg_resources,pluggy,py,pytest,setuptools,six
```

File: tox.ini [hunk 2]
```diff
@@ -103,7 +103,7 @@ source = tox
          {toxworkdir}/pypy*/site-packages/tox
 
 [pytest]
-addopts = -rsxX -vvv --showlocals
+addopts = -rsxX --showlocals
```
Comment (obestwalter) [inline tox.ini]: ok :)

File: setup.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 import io
 import sys
 
-import setuptools
+from setuptools import __version__ as setuptools_version
```
Comment (obestwalter) [inline setup.py]: What is the connection to switching to src/ layout here? Is that some name shadowing thing?

**Comment by nicoddemus** [comment]
Rá, pytest wins the race! https://github.com/pytest-dev/pytest/pull/3513 😸

**Comment by gaborbernat** [comment]
@nicoddemus the reason that's the case is cause I did not figure out yet how to merge coverage files in between interpreters, will probably do it tomorrow when I have some time for that. As I see pytest does not do that 🤔

**Comment by anthrotype** [comment]
you need to run coverage with `--parallel-mode` option and use `[paths] source = ...` in `.coveragerc to list the file paths that should be considered equivalent when combining data:

https://coverage.readthedocs.io/en/coverage-4.5.1/config.html#paths

E.g.:
https://github.com/pyca/cryptography/blob/afdbfb13780fb78e7b277b9de07e7636ba9c5119/tox.ini#L18
https://github.com/pyca/cryptography/blob/afdbfb13780fb78e7b277b9de07e7636ba9c5119/.coveragerc#L7-L11

**Comment by anthrotype** [comment]
@gaborbernat I see. I remember I also tried once to do it with pytest-cov but then gave up and decided to run pytest as `coverage run -m pytest`.

**Comment by gaborbernat** [comment]
@nicoddemus managed to nail down things now; tox is not too far behind! @obestwalter @asottile please review and we can merge :+1:

File: tox.ini [hunk 4]
```diff
@@ -12,47 +12,60 @@ skip_missing_interpreters = true
 
 [testenv]
 description = run the tests with pytest under {basepython}
-setenv = COVERAGE_FILE={toxworkdir}/.coverage.{envname}
+setenv = VIRTUALENV_DOWNLOAD = 0
```
Comment (asottile) [inline tox.ini]: I've usually seen this written as `VIRTUALENV_NO_DOWNLOAD=1`

File: tox.ini [hunk 5]
```diff
@@ -12,47 +12,60 @@ skip_missing_interpreters = true
...
+         PIP_DISABLE_VERSION_CHECK = 1
+         COVERAGE_FILE = {toxworkdir}/.coverage.{envname}
 passenv = http_proxy https_proxy no_proxy SSL_CERT_FILE TOXENV CI TRAVIS TRAVIS_* APPVEYOR APPVEYOR_* CODECOV_*
+deps = setuptools >= 29.0.0
+       setuptools_scm >= 2.0.0, < 3
+       wheel >= 0.31.0
```
Comment (asottile) [inline tox.ini]: you set `--no-download,`, but then ask for an upgrade of `wheel` / `setuptools`? (why not just allow virtualenv to do that for you at this point?)

File: tox.ini [hunk 6]
```diff
@@ -12,47 +12,60 @@ skip_missing_interpreters = true
...
 description = invoke sphinx-build to build the HTML docs and check that all links are valid
 whitelist_externals = sphinx-build
 basepython = python3.6
 extras = docs
+changedir = {toxinidir}
 commands = sphinx-build -d "{toxworkdir}/docs_doctree" doc "{toxworkdir}/docs_out" --color -W -bhtml
 
 
 [testenv:fix-lint]
+description = run static analysis and style check using flake8
```
Comment (asottile) [inline tox.ini]: heh, description isn't exactly right

File: tox.ini [hunk 7]
```diff
@@ -12,47 +12,60 @@ skip_missing_interpreters = true
...
+changedir = {toxinidir}
 commands = pre-commit run --all-files --show-diff-on-failure
            python -c 'print("hint: run {envdir}/bin/pre-commit install to add checks as pre-commit hook")'
 
 
 [testenv:coverage]
-description = combine coverage data and create reports
-deps = coverage
+description = [run locally after tests]: combine coverage data and create report;
+              generates a diff coverage against origin/master (can be changed by setting DIFF_AGAINST env var)
```
Comment (asottile) [inline tox.ini]: should `DIFF_AGAINST` be in `passenv`?

File: tox.ini [hunk 8]
```diff
@@ -12,47 +12,60 @@ skip_missing_interpreters = true
...
+description = run static analysis and style check using flake8
 basepython = python3.6
 passenv = {[testenv]passenv}
           HOMEPATH
           # without PROGRAMDATA cloning using git for Windows will fail with an
           # `error setting certificate verify locations` error
           PROGRAMDATA
 extras = lint
-description = run static analysis and style check using flake8
+changedir = {toxinidir}
```
Comment (asottile) [inline tox.ini]: heh, `pre-comit` will `cd` right back to the repository root, not sure this is useful

File: tox.ini [hunk 9]
```diff
@@ -90,21 +106,21 @@ max-line-length = 99
 ignore = E203, W503
 
 [coverage:run]
-omit = tox/__main__.py
+branch = false
```
Comment (asottile) [inline tox.ini]: branch coverage is useful! please don't turn it off!

File: tox.ini [hunk 10]
```diff
@@ -90,21 +106,21 @@ max-line-length = 99
...
 
 [coverage:paths]
-source = tox
-         {toxworkdir}/*/lib/python*/site-packages/tox
-         {toxworkdir}/*/Lib/site-packages/tox
-         {toxworkdir}/pypy*/site-packages/tox
+source = src/tox
+         .tox/*/lib/python*/site-packages/tox
+         .tox/*/Lib/site-packages/tox
+         .tox/pypy*/site-packages/tox
```
Comment (asottile) [inline tox.ini]: this is by far my least favorite part of moving to a `src` directory, the coverage reports become unreadable :(

we've used [this script](https://github.com/Yelp/venv-update/blob/master/tests/testing/fix_coverage.py) or a derivative of it to try and make this more sane (it combines the coverage data from site-packages over top the local directory)

File: tox.ini (hunk 5 above)
Comment (gaborbernat) [reply tox.ini]: It has to do with the index URL used. By default virtualenv uses PyPi. This way the tox config adheres too whatever the user has configured; e.g. value from ``PIP_INDEX_URL``.

File: tox.ini [hunk 11]
```diff
@@ -12,47 +12,58 @@ skip_missing_interpreters = true
...
 
 [testenv:fix-lint]
+description = format the code base to adhere to our styles, and complain about what we cannot do automatically
 basepython = python3.6
 passenv = {[testenv]passenv}
           HOMEPATH
           # without PROGRAMDATA cloning using git for Windows will fail with an
           # `error setting certificate verify locations` error
           PROGRAMDATA
+          DIFF_AGAINST
```
Comment (asottile) [inline tox.ini]: oops, added this to the wrong testenv, needs to be in the `coverage` testenv

**Comment by gaborbernat** [comment]
things now looks great :+1: so I'll go ahead and merge it. :speak_no_evil: sorry for people with PRs for the merge conflicts

**Pull request merged**